### PR TITLE
Add basic implementation for better-sqlite3 driver

### PR DESCRIPTION
**Pull request by ZIMkaRU**
This PR adds basic implementation for the [better-sqlite3](https://github.com/JoshuaWise/better-sqlite3) driver with an async way to delegate a query into DB using Node.js [worker threads](https://nodejs.org/api/worker_threads.html).
For more details check the `README.md` file
Adds corresponding test coverage

### CODE REVIEW

File: worker/db-worker-actions/index.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+'use strict'
+
+const DB_WORKER_ACTIONS = require('./db-worker-actions.const')
+
+const actionGet = require('./action-get')
+const actionAll = require('./action-all')
+const actionRun = require('./action-run')
+const actionExecPragma = require('./action-exec-pragma')
+
+module.exports = (db, args) => {
```
Comment (ezewer) [inline worker/db-worker-actions/index.js:12]: Here we can do something like: 
```
const DB_WORKER_ACTIONS = { // specify in db-worker-actions.const.js
  CLOSE_DB: closeDb, // Would need to be defined
  GET: actionGet,
  ALL: actionAll,
  RUN: actionRun,
  EXEC_PRAGMA: actionExecPragma
}

module.exports = (db, args) => {
  const { action, sql, params } = args
  if (DB_WORKER_ACTIONS[action]) {
    return DB_WORKER_ACTIONS[action](db, sql, params)
  }
  throw new Error('ERR_ACTION_HAS_NOT_BEEN_FOUND')
}
```

**Review by ezewer — APPROVED** [review]
Small idea, rest looks good to me! 
Really good job here!

File: worker/db-worker-actions/index.js (hunk 1 above)
Comment (ZIMkaRU) [reply worker/db-worker-actions/index.js:12]: really that would look better)
would change

**Review by ezewer — APPROVED** [review]

File: index.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+          }
+          if (code !== 0) {
+            console.error(`worker exited with code ${code}`)
+
+            spawn()
+          }
+        })
+    }
+
+    new Array(this._getRequiredWorkersCount())
```
Comment (robertkowalski) [inline index.js:297]: what happens if a worker dies? is the reference cleaned up? i don't see it

File: index.js [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+          this._queue.length !== 0 &&
+          !jobData.job
+        ) {
+          jobData.job = this._queue.shift()
+          worker.postMessage(jobData.job.message)
+
+          return
+        }
+
+        jobData.timer = setTimeout(poll, 3000)
```
Comment (robertkowalski) [inline index.js:253]: maybe polling just every three seconds is not enough when there is a full queue with 100 jobs?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+      minWorkersCount,
+      maxWorkersCount,
+      unacceptableWalFileSize
+    }
+  }
+
+  _initDb (cb) {
+    const {
+      isSqliteStoredInMemory,
+      isNotWorkerSpawned,
```
Comment (robertkowalski) [inline index.js:71]: naming: i don't understand the setting by its name. it the correct name: "noWorkerThreads"?

File: index.js (hunk 3 above)
Comment (ezewer) [reply index.js:253]: The 3 seconds rule is just for when there are no jobs in queue

File: index.js [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+          return
+        }
+
+        const dbDir = path.dirname(dbPath)
+
+        fs.access(dbDir, fs.constants.W_OK, (err) => {
+          if (err && err.code === 'ENOENT') {
+            const msg = `the directory ${dbDir} does not exist, please create`
+
+            return next(new Error(msg))
```
Comment (robertkowalski) [inline index.js:134]: `ERR_NO_DIRECTORY`

File: worker/worker-factory.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      name: err.name,
+      message: err.message,
+      stack: err.stack,
+      isError: true
+    })
+  }
+
+  const db = _connect(workerData)
+
+  parentPort.on('message', (args) => {
```
Comment (robertkowalski) [inline worker/worker-factory.js:50]: did you do some benchmarking, for lets say 500k writes and 500k reads and compared it with the our other fac? switching the context and keeping the workers also has some overhead and we want to make sure that we improve perf

File: index.js (hunk 3 above)
Comment (ZIMkaRU) [reply index.js:253]: So we are pulling a job by a worker when adding a task by `asyncQuery` method or when a worker will finish a current taken task `.on('message', process)` or as Ezequiel said

File: index.js (hunk 4 above)
Comment (ZIMkaRU) [reply index.js:71]: yes, it's the setting to avoid spawning any worker threads
would change to that naming

File: index.js (hunk 5 above)
Comment (ZIMkaRU) [reply index.js:134]: sure)
just wanted to do the error message similar message such as it did in [bfx-facs-db-sqlite](https://github.com/bitfinexcom/bfx-facs-db-sqlite/blob/master/index.js#L47)

File: worker/worker-factory.js (hunk 6 above)
Comment (ZIMkaRU) [reply worker/worker-factory.js:50]: Currently, we are in progress in reports tool, doing a migration to this driver and also will leave previously db implementation to have possible to benchmark
The final result we can see on reports tool with adding some test
Some benchmark is already done in better-sqlite3 (but without worker threads) https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/benchmark.md

**Pull request merged**